org.eclipse.xtext/src/org/eclipse/xtext/validation/LocalUniqueNameContext.java

```diff
@@ -0,0 +1,189 @@
+				index.merge(name, description, (p, n)->{
+					if (p instanceof List<?>) {
+						@SuppressWarnings("unchecked")
+						List<Object> list = (List<Object>) p;
+						list.add(n);
+						return list;
+					}
+					List<Object> list = new ArrayList<>();
+					list.add(p);
+					list.add(n);
+					return list;
+				});
+			}
+		}
+		this.objectsToValidate = objectsToValidate;
+		this.index = index;
+		this.ci = ci;
+	}
+
+	public LocalUniqueNameContext(List<? extends EObject> objects, CancelIndicator ci) {
+		this(objects, LocalUniqueNameContext::tryGetName, ci);
+	}
+
+	@Override
+	public boolean isUnique() {
+		return objectsToValidate.size() == index.size();
+	}
+
+	@Override
+	public final boolean isCaseSensitive(EObject candidate, EClass clusterType) {
+		return true;
+	}
+
+	@Override
+	public ISelectable getValidationScope(IEObjectDescription description, EClass clusterType) {
+		return this;
+	}
+
+	@Override
+	public List<IEObjectDescription> getObjectsToValidate() {
+		return objectsToValidate;
+	}
+
+	@Override
+	public CancelIndicator cancelIndicator() {
+		return ci;
+	}
+
+	@Override
+	public Iterable<IEObjectDescription> getExportedObjects() {
+		return getObjectsToValidate();
+	}
+
+	@Override
+	public boolean isEmpty() {
+		return index.isEmpty();
+	}
+
+	@Override
+	public Iterable<IEObjectDescription> getExportedObjects(EClass type, QualifiedName name, boolean ignoreCase) {
+		Preconditions.checkArgument(!ignoreCase);
+		Object result = index.get(name.getFirstSegment());
+		if (result == null) {
+			return Collections.emptyList();
+		}
+		if (result instanceof IEObjectDescription) {
+			if (EcoreUtil2.isAssignableFrom(type, ((IEObjectDescription) result).getEClass())) {
+				return Collections.singletonList((IEObjectDescription)result);
+			}
+			return Collections.emptyList();
+		}
+		@SuppressWarnings("unchecked")
+		List<IEObjectDescription> casted = (List<IEObjectDescription>) result;
+		if (type == EcorePackage.Literals.EOBJECT) {
+			return casted;
+		}
+		return FluentIterable.from(casted).filter(it -> EcoreUtil2.isAssignableFrom(type, it.getEClass()));
+	}
+
+	@Override
+	public Iterable<IEObjectDescription> getExportedObjectsByObject(EObject object) {
+		return FluentIterable.from(objectsToValidate).filter(it -> object == it.getEObjectOrProxy());
+	}
+
+	@Override
+	public Iterable<IEObjectDescription> getExportedObjectsByType(EClass type) {
+		return FluentIterable.from(objectsToValidate).filter(it -> EcoreUtil2.isAssignableFrom(type, it.getEClass()));
+	}
+
+	/**
+	 * Read the value of an EAttribute 'name', if present. Returns null of the given object is null or does not have a
+	 * name.
+	 */
+	public static String tryGetName(EObject obj) {
+		if (obj == null) {
+			return null;
+		}
+		EStructuralFeature name = obj.eClass().getEStructuralFeature("name");
+		if (name != null && name.getEType() == EcorePackage.Literals.ESTRING) {
+			return (String) obj.eGet(name);
+		}
+		return null;
+	}
```

Comment:
Could we use `SimpleAttributeResolver.NAME_RESOLVER.getAttribute(obj);` here, as done in the [AbstractRenameStrategy](https://github.com/eclipse/xtext-eclipse/blob/106b597f755f63be3ea81c88865ce38c0b4cf343/org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/refactoring/impl/AbstractRenameStrategy.java#L86-L88) class?

Reply:
The SimpleAttributeResolver uses all sorts of weird caching which requires locking etc. I'd rather like to get rid of that one.

Reply:
It even installs adapters on the objects which is really worrisome.

Reply:
I see, than it is ok for me.